File: java/example/src/main/java/CacheCredTest.java
import io.antidrift.zeromcp.*;

import com.google.gson.*;
import java.nio.file.*;
import java.util.*;

public class CacheCredTest {

    private static String readTokenFromFile(String path) {
        try {
            var text = Files.readString(Path.of(path));
            var obj = new Gson().fromJson(text, JsonObject.class);
            var el = obj.get("token");
            return el != null && !el.isJsonNull() ? el.getAsString() : null;
        } catch (Exception e) {
            return null;
        }
    }

    public static void main(String[] args) {
        var configPath = System.getenv("ZEROMCP_CONFIG");
        if (configPath == null) configPath = "zeromcp.config.json";

        // Parse config to get credentials.tokenstore.file and cache_credentials.
        String credFile = "";
        boolean cacheCredentials = true;
        try {
            var text = Files.readString(Path.of(configPath));
            var cfg = new Gson().fromJson(text, JsonObject.class);
            if (cfg != null) {
                var creds = cfg.getAsJsonObject("credentials");
                if (creds != null) {
                    var tokenstore = creds.getAsJsonObject("tokenstore");
                    if (tokenstore != null && tokenstore.has("file")) {
                        credFile = tokenstore.get("file").getAsString();
                    }
                }
                if (cfg.has("cache_credentials")) {
                    cacheCredentials = cfg.get("cache_credentials").getAsBoolean();
                }
            }
        } catch (Exception ignored) {}

        final String finalCredFile = credFile;
        // When caching is enabled, read once at startup.
        final String[] cachedToken = { null };
        final boolean[] cached = { false };
        final boolean doCache = cacheCredentials;

        var server = new ZeroMcp();

        server.tool("tokenstore_check", Tool.builder()
            .description("Return the current token from credentials")
            .execute((a, ctx) -> {
                String token;
                if (doCache) {
                    if (!cached[0]) {
                        cachedToken[0] = readTokenFromFile(finalCredFile);
                        cached[0] = true;
                    }
                    token = cachedToken[0];
                } else {
                    token = readTokenFromFile(finalCredFile);
                }
                var result = new HashMap<String, Object>();
                result.put("token", token);
                return result;
            })
            .build());

        server.serve();
    }
}

File: rust/examples/cache_cred_test.rs
use serde_json::Value;
use std::sync::{Arc, Mutex};
use zeromcp::{Ctx, Input, Permissions, Server, Tool};

fn read_token_from_file(path: &str) -> Option<String> {
    let text = std::fs::read_to_string(path).ok()?;
    let v: Value = serde_json::from_str(&text).ok()?;
    v.get("token")?.as_str().map(|s| s.to_string())
}

#[tokio::main]
async fn main() {
    let config_path = std::env::var("ZEROMCP_CONFIG")
        .unwrap_or_else(|_| "zeromcp.config.json".to_string());

    // Parse config JSON to get credentials.tokenstore.file and cache_credentials flag.
    let raw = std::fs::read_to_string(&config_path).unwrap_or_else(|_| "{}".to_string());
    let config_json: Value = serde_json::from_str(&raw).unwrap_or(Value::Object(Default::default()));

    let cred_file = config_json
        .pointer("/credentials/tokenstore/file")
        .and_then(|v| v.as_str())
        .unwrap_or("")
        .to_string();

    let cache_credentials = config_json
        .get("cache_credentials")
        .and_then(|v| v.as_bool())
        .unwrap_or(true);

    // When caching is enabled, resolve once at startup.
    let cached: Arc<Mutex<Option<Option<String>>>> = Arc::new(Mutex::new(None));

    let cred_file_clone = cred_file.clone();
    let cached_clone = cached.clone();

    let mut server = Server::from_config(&config_path);

    server.tool(
        "tokenstore_check",
        Tool {
            description: "Return the current token from credentials".to_string(),
            input: Input::new(),
            permissions: Permissions::default(),
            execute: Box::new(move |_args: Value, _ctx: Ctx| {
                let path = cred_file_clone.clone();
                let cache = cached_clone.clone();
                Box::pin(async move {
                    let token = if cache_credentials {
                        let mut lock = cache.lock().unwrap();
                        if lock.is_none() {
                            *lock = Some(read_token_from_file(&path));
                        }
                        lock.clone().unwrap()
                    } else {
                        read_token_from_file(&path)
                    };
                    Ok(serde_json::json!({ "token": token }))
                })
            }),
            cached_schema: Default::default(),
        },
    );

    server.serve().await;
}
